### PR TITLE
Enable standalone shader cache creation for relocatable shaders

### PR DESCRIPTION
**Pull request by kuhar**
This PR adds two new lit feature tests: "assertions" and
"llpc-shader-cache" that can be queried in "REQUIRES", "UNSUPPORTED",
 and "XFAIL".

All relocatable shaders tests are moved into their own directory in
llpc/tests/shaderdb.

The cache creation code is restarted from #455.

This PR depends on #633 -- please review the latest commit only. Testing will be added in a separate PR after #634 lands.

### CODE REVIEW

**Comment by amdvlk-admin** [comment]
Can one of the admins verify this patch?

File: llpc/tool/amdllpc.cpp [hunk 1]
```diff
@@ -1611,9 +1612,33 @@ int main(int argc, char *argv[]) {
   if (isFailure())
     return onFailure();
 
-  // The first input file is a pipeline file or LLVM IR file. Assume they all are, and compile each one
-  // separately but in the same context.
-  if (isPipelineInfoFile(expandedInputFiles[0]) || isLlvmIrFile(expandedInputFiles[0])) {
+  // Build relocatable shader cache. This requires all inputs to be .pipe files.
```
Comment (s-perron) [inline llpc/tool/amdllpc.cpp:1615]: We should mention that this is still a work in progress, and eventually we will want to accept plain spir-v files.

Comment (kuhar) [reply llpc/tool/amdllpc.cpp]: Done

**Comment by JaxLinAMD** [comment]
retest this please

**Comment by kuhar** [comment]
Ping.

**Comment by kuhar** [comment]
Rebased.

File: llpc/context/llpcCompiler.cpp [hunk 2]
```diff
@@ -110,6 +110,12 @@ opt<int> RelocatableShaderElfLimit("relocatable-shader-elf-limit",
                                             "relocatable shader ELF.  -1 means unlimited."),
                                    init(-1));
 
+// -build-relocatable-shader-cache: Populates the shader cache with relocatable shader variants.
+opt<bool> BuildShaderCache("build-shader-cache",
+                           cl::desc("[WiP] Populates shader cache with relocatable shader variants."
```
Comment (nhaehnle) [inline llpc/context/llpcCompiler.cpp:115]: Please capitalize fully as WIP.

File: llpc/context/llpcCompiler.cpp [hunk 3]
```diff
@@ -743,13 +750,13 @@ Result Compiler::buildPipelineWithRelocatableElf(Context *context, ArrayRef<cons
       auto pipelineInfo = reinterpret_cast<const GraphicsPipelineBuildInfo *>(context->getPipelineBuildInfo());
       cacheHash = PipelineDumper::generateHashForGraphicsPipeline(pipelineInfo, true, true, stage);
 #if LLPC_CLIENT_INTERFACE_MAJOR_VERSION < 38 || LLPC_ENABLE_SHADER_CACHE
-      userShaderCache = pipelineInfo->shaderCache;
+      userShaderCache = static_cast<IShaderCache *>(static_cast<void *>(pipelineInfo->shaderCache));
```
Comment (nhaehnle) [inline llpc/context/llpcCompiler.cpp:753]: Why is this needed, here and elsewhere?

File: llpc/context/llpcCompiler.cpp [hunk 4]
```diff
@@ -1485,14 +1498,26 @@ Result Compiler::BuildComputePipeline(const ComputePipelineBuildInfo *pipelineIn
...
+                                       cl::EnablePipelineDump.ArgStr,
+                                       cl::ShaderCacheFileDir.ArgStr,
+                                       cl::ShaderCacheMode.ArgStr,
+                                       cl::EnableOuts.ArgStr,
+                                       cl::EnableErrs.ArgStr,
+                                       cl::LogFileDbgs.ArgStr,
+                                       cl::LogFileOuts.ArgStr,
+                                       cl::ExecutableName.ArgStr,
+                                       cl::BuildShaderCache.ArgStr,
+                                       "o"};
```
Comment (nhaehnle) [inline llpc/context/llpcCompiler.cpp:1511]: FWIW, I consider combining a significant style change with a semantic change like this a bit rude. It'd be nicer to have that split into separate commits. I can live with it here, though.

File: llpc/context/llpcCompiler.cpp (hunk 3 above)
Comment (kuhar) [reply llpc/context/llpcCompiler.cpp:753]: The forward declaration was moved into Vkgc, while the class/struct definition remained in Llpc. I think this was an innocuous breakage that nobody noticed since this code is not currently being used with the current llpc client interface. I suggest a proper fix can follow once this code is built and tested.

File: llpc/context/llpcCompiler.cpp (hunk 4 above)
Comment (kuhar) [reply llpc/context/llpcCompiler.cpp:1511]: Thanks for the suggestion. Do I understand correctly that you propose to first turn off clang-format here and add the extra options, and then remove the formatting exception and reformat the initializer in a subsequent commit?

**Comment by kuhar** [comment]
Rebased.

**Comment by kuhar** [comment]
Ping. @nhaehnle @amdrexu  @JaxLinAMD  @trenouf

**Comment by amdrexu** [comment]
@kuhar I will keep it open today. If no one has further comments, will approve tomorrow morning. For another your change, Nicolai requested changes and blocked it. So this needs Nicolai's reapproval. Otherwise, it is still in "blocked" status and no one can merge it.

File: llpc/context/llpcCompiler.cpp (hunk 3 above)
Comment (nhaehnle) [reply llpc/context/llpcCompiler.cpp]: Ah, one more argument against preprocessor use... I think it can be forgiven for now ;)

**Comment by JaxLinAMD** [comment]
retest this please

**Comment by amdvlk-admin** [comment]
### Test summary for commit 72e811357a62dd685831383857e384c5530a07dd
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master b601c88aeca7a7b08becb3d32709de383c8ee428</li>
<li>METROHASH: amd-master 2b6fee002db6cc92345b02aeee963ebaaf4c0e2f</li>
<li>PAL: dev ddcaa1009ecde219613042a5a53c6d3752bf4b9d</li>
<li>SPVGEN: dev d4817ab957f79762da83b7fc1d3f2816dbdc13fd</li>
<li>XGL: dev 2ecd2e14e2ae4f2712df4a2220e2dd36265c66d1</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev c57487b76215ee74b1038039f47b210a28cad65e</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 959/959 (100.0%)</li>
<li>Failed: 0/959 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix)</summary><ul>


<details open><summary>Ubuntu 18.04, Ellesmere</summary><ul>

<li>Passed: 103143/103143 (100.0%)</li>
<li>Failed: 0/103143 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Navi10</summary><ul>

<li>Passed: 103199/103199 (100.0%)</li>
<li>Failed: 0/103199 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 7.6, Vega10</summary><ul>

<li>Passed: 103199/103199 (100.0%)</li>
<li>Failed: 0/103199 (0.0%)</li>
</ul></details>


</ul></details>
